include/state-observation/tools/rigid-body-kinematics.hxx

```diff
@@ -2113,63 +2385,1701 @@ inline const Kinematics & Kinematics::update_deprecated(const Kinematics & newVa
+  }
+
+  if(flagAngAcc)
+  {
+    angAcc.set().setZero();
+  }
+
+  return *this;
+}
+
+inline LocalKinematics & LocalKinematics::setZero(LocalKinematics::Flags::Byte flags)
+{
+  return setZero<Quaternion>(flags);
+}
+
+inline const LocalKinematics & LocalKinematics::integrateRungeKutta4(
+    double dt,
+    RecursiveAccelerationFunctorBase & accelerationFunctor)
+{
+  BOOST_ASSERT((position.isSet() && orientation.isSet() && linVel.isSet() && angVel.isSet())
+               && "The kinematics of the LocalKinematics object are not entirely set");
+  BOOST_ASSERT((linAcc.isSet() && angAcc.isSet())
+               && "The accelerations of the LocalKinematics object are not entirely set");
+  LocalKinematics y234;
+
+  LocalKinematics::Derivative k1; // (velocities + accelerations) that have to be precomputed
+  LocalKinematics::Derivative k2;
+  LocalKinematics::Derivative k3;
+  LocalKinematics::Derivative k4;
+
+  k1 = *this;
+
+  y234.position = position() + 0.5 * dt * k1.positionDot;
+  y234.linVel = linVel() + 0.5 * dt * k1.linVelDot;
+
+  y234.orientation = orientation;
+  y234.orientation.integrateRightSide(0.5 * dt * k1.angVel);
+  y234.angVel = angVel() + 0.5 * dt * k1.angAcc;
+
+  accelerationFunctor.computeRecursiveLocalAccelerations_(
+      y234); // computation of k2 (the acceleration part) in the global frame
+
+  // conversion of k2 to the local frame in order to make them compatible with the LocalKinematics object
+  k2 = y234;
+
+  y234.position = position() + 0.5 * dt * k2.positionDot;
+  y234.linVel = linVel() + 0.5 * dt * k2.linVelDot;
+
+  y234.orientation = orientation;
+  y234.orientation.integrateRightSide(0.5 * dt * k2.angVel);
+  y234.angVel = angVel() + 0.5 * dt * k2.angAcc;
+
+  accelerationFunctor.computeRecursiveLocalAccelerations_(
+      y234); // computation of k3 (the acceleration part) in the global frame
+
+  // conversion of k3 to the local frame in order to make them compatible with the LocalKinematics object
+  k3 = y234;
+
+  y234.position = position() + dt * k3.positionDot;
+  y234.linVel = linVel() + dt * k3.linVelDot;
+
+  y234.orientation = orientation;
+  y234.orientation.integrateRightSide(dt * k3.angVel);
+  y234.angVel = angVel() + dt * k3.angAcc;
+
+  accelerationFunctor.computeRecursiveLocalAccelerations_(
+      y234); // computation of k4 (the acceleration part) in the global frame
+
+  // conversion of k4 to the local frame in order to make them compatible with the LocalKinematics object
+  k4 = y234;
+
+  position() += dt / 6 * (k1.positionDot + 2 * k2.positionDot + 2 * k3.positionDot + k4.positionDot);
+  orientation.integrateRightSide(dt / 6 * (k1.angVel + 2 * k2.angVel + 2 * k3.angVel + k4.angVel));
+  linVel() += dt / 6 * (k1.linVelDot + 2 * k2.linVelDot + 2 * k3.linVelDot + k4.linVelDot);
+  angVel() += dt / 6 * (k1.angAcc + 2 * k2.angAcc + 2 * k3.angAcc + k4.angAcc);
+
+  return *this;
+}
+
+inline const LocalKinematics & LocalKinematics::integrate(double dt)
```

Comment:
Won't it work if AreSet is a Flags object? Need to check...
Maybe turn AreSet into an inline function too

Reply:
> Won't it work if AreSet is a Flags object?

Not really but it was a `Flags::Byte` yes :)

> Maybe turn AreSet into an inline function too

Definitely. This was meant to show the technique but if it's used in a few places it can be integrated more cleanly.

Also, I have realized there's some issue with the code I suggested:
- in the first switch statement we should use `areSet & ~Orientation`
- in the second switch statement we should use `areSet & ~(Position|AngAcc)`
- in the third switch statement we should use `areSet & ~(Position | LinVel | LinAcc)`

Reply:
Added this change in commit 44eb405120edb96f1a3098b01ab4686c4b197276: fixed wrong velocity integration
In order for linVel to be correctly computed, the switch statement had to be `switch(areSet & ~(Position | Orientation | AngAcc))`